### PR TITLE
feat(eslint-plugin): [no-floating-promises] error on logical expression

### LINKED ISSUE #6343: Bug: [no-floating-promises] Not marking as error when combined with a logical operator variable
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.9.3&sourceType=module&code=IYZwngdgxgBAZgV2gFwJYHsL3egFAShgG8BYAKBhikxGRgFswAFAJ3XtRAFMYBeGVu05cAdCy4h0AGwBuXXMhYIu+ANzlK1CLSqYAJqjSY+MRcvVkNuiAaNYAZPYbM2HbhYC+QA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oDN4OBDfMwDmtYtA4BbSshTooiaBOiQANOGx4ipZBWp0GzNp1rTksWQBMxEs3IyRFyyBoC+IV0A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
async function foo() {
  const myPromise = Promise.resolve(true);
  const condition = true;

  condition && myPromise;
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-floating-promises": "error",
    "@typescript-eslint/no-misused-promises": "error",
  },
};
```

### Expected Result

This promise should be marked as `floating-promises` or `misused-promise`.

```
// Wrong usage
 condition && myPromise;

// Correct usage
 condition && await myPromise;
```

### Actual Result

This promise is not marked as floating/misused.

### Additional Info

Similar issue was closed: https://github.com/typescript-eslint/typescript-eslint/issues/2544

### PR DESCRIPTION
**Pull request by islandryu**
## PR Checklist

- [x] Addresses an existing open issue: fixes #6343
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

Promise in logical expression now also throws an error.

### CODE REVIEW

File: packages/eslint-plugin/tests/rules/no-floating-promises.test.ts [hunk 1]
```diff
@@ -1117,5 +1168,264 @@ async function test() {
...
+          messageId: 'floatingVoid',
+          suggestions: [
+            {
+              messageId: 'floatingFixVoid',
+              output: `
+async function foo() {
+  const myPromise = async () => void 0;
+  const condition = true;
+
+  void (condition && myPromise());
```
Comment (islandryu) [inline packages/eslint-plugin/tests/rules/no-floating-promises.test.ts:1251]: The entire logical expression is enclosed in parentheses as in return-await.
https://typescript-eslint.io/play/#ts=4.9.3&sourceType=module&code=IYZwngdgxgBAZgV2gFwJYHsIwEbAE4AUAlDAN4wC+AsAFCiSyIoZZzrrFm0wzJ5hcaPHngCmyBHixIAJqLioIomTAA+qnPmIBubpRhRgyKAAsylWhSA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6aRfWaJ2gQwDufSvnRRE0aAHtokcGAC+IBUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

File: packages/eslint-plugin/src/rules/no-floating-promises.ts [hunk 2]
```diff
@@ -136,16 +148,13 @@ export default util.createRule<Options, MessageId>({
...
         ? node.operatorToken.kind
         : ts.SyntaxKind.Unknown;
       const nodePrecedence = util.getOperatorPrecedence(node.kind, operator);
-      const awaitPrecedence = util.getOperatorPrecedence(
-        ts.SyntaxKind.AwaitExpression,
-        ts.SyntaxKind.Unknown,
-      );
-      return nodePrecedence > awaitPrecedence;
+      const unaryPrecedence = OperatorPrecedence.Unary;
+      return nodePrecedence > unaryPrecedence;
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-floating-promises.ts:157]: Nit: I don't see a strong reason to make a variable here, given that its name is a subset of its value's words?

```suggestion
      return nodePrecedence > OperatorPrecedence.Unary;
```

**Review by JoshuaKGoldberg — DISMISSED** [review]
LGTM, thanks! Just fix the build failure and 🚀 let's ship.

I was on the fence as to whether the suggestions should put the `await`/`void` immediately in front of the specific promise, rather than outside the logical expression (i.e. `a || (await b)` rather than `await (a || b)`. But since it's just suggestions, not auto-fixing, I think this is fine to keep the rule simpler.

File: packages/eslint-plugin/src/rules/no-floating-promises.ts (hunk 2 above)
Comment (islandryu) [reply packages/eslint-plugin/src/rules/no-floating-promises.ts]: You are right. Fixed!

**Comment by islandryu** [comment]
Build is cured!

File: packages/eslint-plugin/src/rules/no-floating-promises.ts [hunk 3]
```diff
@@ -88,10 +89,21 @@ export default util.createRule<Options, MessageId>({
               suggest: [
                 {
                   messageId: 'floatingFixVoid',
-                  fix(fixer): TSESLint.RuleFix {
-                    let code = sourceCode.getText(node);
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-floating-promises.ts:92]: 🤔 I'm surprised the build isn't failing - but `sourceCode` is indeed no longer used. Could you remove the import please? Then we can merge!

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
One teeeeny thing

File: packages/eslint-plugin/src/rules/no-floating-promises.ts (hunk 3 above)
Comment (islandryu) [reply packages/eslint-plugin/src/rules/no-floating-promises.ts:92]: Thank you!
I had overlooked that.

Comment (islandryu) [reply packages/eslint-plugin/src/rules/no-floating-promises.ts:92]: I don't know what caused the build to fail either, but it was cured when I merged the main branch.

**Review by JoshuaKGoldberg — APPROVED** [review]
🔥

**Pull request merged**